cpp/src/arrow/util/async_util_test.cc

```diff
@@ -0,0 +1,150 @@
+}
+
+TEST(AsyncDestroyable, MakeShared) {
+  TestAsyncDestroyable([](Future<> gate, bool* destroyed) {
+    return MakeSharedAsync<GatingDestroyable>(gate, destroyed);
+  });
+}
+
+TEST(AsyncDestroyable, MakeUnique) {
+  Future<> gate = Future<>::Make();
+  bool destroyed = false;
+  bool on_closed = false;
+  {
+    std::unique_ptr<GatingDestroyable, DestroyingDeleter<GatingDestroyable>> obj =
+        MakeUniqueAsync<GatingDestroyable>(gate, &destroyed);
+    obj->on_closed().AddCallback([&](const Status& st) { on_closed = true; });
+    ASSERT_FALSE(destroyed);
+  }
+  ASSERT_FALSE(destroyed);
+  ASSERT_FALSE(on_closed);
+  gate.MarkFinished();
+  ASSERT_TRUE(destroyed);
+  ASSERT_TRUE(on_closed);
+}
+
+TEST(AsyncTaskGroup, Basic) {
+  AsyncTaskGroup task_group;
+  Future<> fut1 = Future<>::Make();
+  Future<> fut2 = Future<>::Make();
+  ASSERT_OK(task_group.AddTask(fut1));
+  ASSERT_OK(task_group.AddTask(fut2));
+  Future<> all_done = task_group.WaitForTasksToFinish();
+  AssertNotFinished(all_done);
+  fut1.MarkFinished();
+  AssertNotFinished(all_done);
+  fut2.MarkFinished();
+  ASSERT_FINISHES_OK(all_done);
+}
+
+TEST(AsyncTaskGroup, NoTasks) {
+  AsyncTaskGroup task_group;
+  ASSERT_FINISHES_OK(task_group.WaitForTasksToFinish());
+}
+
+TEST(AsyncTaskGroup, AddAfterDone) {
+  AsyncTaskGroup task_group;
+  ASSERT_FINISHES_OK(task_group.WaitForTasksToFinish());
+  ASSERT_RAISES(Invalid, task_group.AddTask(Future<>::Make()));
+}
+
+TEST(AsyncTaskGroup, AddAfterWaitButBeforeFinish) {
+  AsyncTaskGroup task_group;
+  Future<> task_one = Future<>::Make();
+  ASSERT_OK(task_group.AddTask(task_one));
+  Future<> finish_fut = task_group.WaitForTasksToFinish();
+  AssertNotFinished(finish_fut);
+  Future<> task_two = Future<>::Make();
+  ASSERT_OK(task_group.AddTask(task_two));
+  AssertNotFinished(finish_fut);
+  task_one.MarkFinished();
+  AssertNotFinished(finish_fut);
+  task_two.MarkFinished();
+  AssertFinished(finish_fut);
+  ASSERT_FINISHES_OK(finish_fut);
+}
+
+TEST(AsyncTaskGroup, Error) {
+  AsyncTaskGroup task_group;
+  Future<> failed_task = Future<>::MakeFinished(Status::Invalid("XYZ"));
+  ASSERT_OK(task_group.AddTask(failed_task));
+  ASSERT_FINISHES_AND_RAISES(Invalid, task_group.WaitForTasksToFinish());
+}
+
+TEST(AsyncTaskGroup, TaskFinishesAfterError) {
+  AsyncTaskGroup task_group;
+  Future<> fut1 = Future<>::Make();
+  ASSERT_OK(task_group.AddTask(fut1));
+  ASSERT_OK(task_group.AddTask(Future<>::MakeFinished(Status::Invalid("XYZ"))));
+  Future<> finished_fut = task_group.WaitForTasksToFinish();
+  AssertNotFinished(finished_fut);
+  fut1.MarkFinished();
+  ASSERT_FINISHES_AND_RAISES(Invalid, finished_fut);
+}
+
+TEST(AsyncTaskGroup, AddAfterFailed) {
+  AsyncTaskGroup task_group;
+  ASSERT_OK(task_group.AddTask(Future<>::MakeFinished(Status::Invalid("XYZ"))));
+  ASSERT_RAISES(Invalid, task_group.AddTask(Future<>::Make()));
+  ASSERT_FINISHES_AND_RAISES(Invalid, task_group.WaitForTasksToFinish());
+}
+
```

Comment:
Since AddTask special-cases adding completed futures, maybe for completeness we should have a test of marking a future failed after it's been added to the group as well.

Reply:
I added a FailsAfterAdd test.